### PR TITLE
support project-scoped gitlab.com runners registration

### PR DESCRIPTION
**Pull request by drzraf**
##### SUMMARY
support project-scoped gitlab.com runners registration

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
gitlab_runner

##### ADDITIONAL INFORMATION
Reroll of https://github.com/ansible/ansible/pull/58840

The current module register runners on global scope (in particular, trying to fetch instance-wide listing of runners) what GitLab.com obviously forbids.
But runners can be registered by-project (and by-group although GitLab API does not support it yet).
By opting for project-scope, then listing and idempotent registration are allowed.
Using a (existing) project is also a convenient way to register a group runner (by providing a group registration token). This avoid the failure of requesting instance-wide runners at GitLab.com

This PR also set debug when using extra-verbose options which is quite handy, especially for such a new module.

### CODE REVIEW

**Comment by felixfontein** [comment]
You need a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

File: plugins/modules/source_control/gitlab/gitlab_runner.py [hunk 1]
```diff
@@ -39,6 +39,11 @@
       - Your private token to interact with the GitLab API.
     required: True
     type: str
+  project:
+    description:
+      - Id or Full path of the project in the form of group/name.
+    required: true
+    type: str
```
Comment (Andersson007) [inline plugins/modules/source_control/gitlab/gitlab_runner.py:46]:
```suggestion
    type: str
    version_added: '1.3.0'
```

**Review by Andersson007 — COMMENTED** [review]
we must have a changelog fragment as felixfontein mentioned
also please add an example for the option

File: changelogs/fragments/634-gitlab_project_runners.yaml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- gitlab_runner - support project-scoped gitlab.com runners registration (https://github.com/ansible-collections/community.general/pull/634)
```
Comment (felixfontein) [inline changelogs/fragments/634-gitlab_project_runners.yaml:2]:
```suggestion
- gitlab_runner - support project-scoped gitlab.com runners registration (https://github.com/ansible-collections/community.general/pull/634).
```

File: plugins/modules/source_control/gitlab/gitlab_runner.py [hunk 3]
```diff
@@ -313,13 +333,23 @@ def main():
...
     maximum_timeout = module.params['maximum_timeout']
     registration_token = module.params['registration_token']
+    project = module.params['project']
 
     if not HAS_GITLAB_PACKAGE:
         module.fail_json(msg=missing_required_lib("python-gitlab"), exception=GITLAB_IMP_ERR)
 
     gitlab_instance = gitlabAuthentication(module)
+    if module._verbosity >= 2:
+        gitlab_instance.enable_debug()
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_runner.py:343]: What's the effect of this?

File: plugins/modules/source_control/gitlab/gitlab_runner.py [hunk 4]
```diff
@@ -39,6 +39,12 @@
       - Your private token to interact with the GitLab API.
     required: True
     type: str
+  project:
+    description:
+      - Id or Full path of the project in the form of group/name.
```
Comment (Andersson007) [inline plugins/modules/source_control/gitlab/gitlab_runner.py:44]:
```suggestion
      - Id or full path of the project in the form of group/name.
```

File: plugins/modules/source_control/gitlab/gitlab_runner.py [hunk 5]
```diff
@@ -115,6 +121,15 @@
     api_token: "{{ access_token }}"
     description: Docker Machine t1
     state: absent
+
+- name: "Register runner for a specific project"
```
Comment (Andersson007) [inline plugins/modules/source_control/gitlab/gitlab_runner.py:125]:
```suggestion
- name: Register runner for a specific project
```

**Comment by felixfontein** [comment]
@drzraf are you still working on this?

needs_info

**Comment by drzraf** [comment]
Oh, sure. Went out of the radar, will try to address comments soon.

**Comment by felixfontein** [comment]
@drzraf ping

needs_info

File: plugins/modules/source_control/gitlab/gitlab_runner.py (hunk 3 above)
Comment (drzraf) [reply plugins/modules/source_control/gitlab/gitlab_runner.py:372]: It provides actually useful debug output in case user requested it.

Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_runner.py:372]: But where does the debug output go? For using `module._verbosity` to make sense, it needs to be sent back to the controller.

File: plugins/modules/source_control/gitlab/gitlab_runner.py [hunk 6]
```diff
@@ -38,6 +38,12 @@
     description:
       - Your private token to interact with the GitLab API.
     type: str
+  project:
+    description:
+      - Id or full path of the project in the form of group/name.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_runner.py:43]:
```suggestion
      - ID or full path of the project in the form of group/name.
```

File: plugins/modules/source_control/gitlab/gitlab_runner.py [hunk 7]
```diff
@@ -38,6 +38,12 @@
...
+    required: true
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_runner.py:44]:
```suggestion
```

File: plugins/modules/source_control/gitlab/gitlab_runner.py (hunk 3 above)
Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_runner.py:372]: @drzraf ping ^

**Pull request closed without merging**